Build failure: scalapack (on aarch64-darwin)

### Steps To Reproduce

Steps to reproduce the behavior:

Pinned to current unstable:

```console
$ nix build github:nixos/nixpkgs/ecd9a5753a5078132dc9bfb95902abecbed31eef#scalapack
error: builder for '/nix/store/cdh4zm4r2n9zjx2p06wvypmz16g04jn7-scalapack-2.2.1.drv' failed with exit code 2;
       last 10 log lines:
       > /tmp/nix-build-scalapack-2.2.1.drv-0/source/BLACS/SRC/dgamx2d_.c:368:10: error: call to undeclared function 'BI_TransDist'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       >          BI_TransDist(ctxt, tscope, Mpval(m), Mpval(n), rA, cA, tldia,
       >          ^
       > /tmp/nix-build-scalapack-2.2.1.drv-0/source/BLACS/SRC/dgamx2d_.c:373:30: error: call to undeclared function 'BI_dvmcopy'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
       >       if (bp != &BI_AuxBuff) BI_dvmcopy(Mpval(m), Mpval(n), A, tlda, bp->Buff);
       >                              ^
       > 8 errors generated.
       > make[2]: *** [CMakeFiles/scalapack.dir/build.make:734: CMakeFiles/scalapack.dir/BLACS/SRC/dgamx2d_.c.o] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:1234: CMakeFiles/scalapack.dir/all] Error 2
       > make: *** [Makefile:166: all] Error 2
       For full logs, run 'nix log /nix/store/cdh4zm4r2n9zjx2p06wvypmz16g04jn7-scalapack-2.2.1.drv'.
```

### Build log

https://gist.github.com/n8henrie/6c7e1c6a63067564fae006207847b902

### Additional context

Related? https://github.com/open-mpi/ompi/issues/6114

### Notify maintainers

@costrouc @markuskowa @gdinh

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.5.0, macOS 14.5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.19.4`
 - channels(n8henrie): `""`
 - channels(root): `""`
 - nixpkgs: `/nix/store/rash1v5mpf4hsz90dap6q38djl7xlnda-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/science/math/scalapack/default.nix
+++ b/pkgs/development/libraries/science/math/scalapack/default.nix
@@ -52,10 +52,11 @@ stdenv.mkDerivation rec {
       -DLAPACK_LIBRARIES="-llapack"
       -DBLAS_LIBRARIES="-lblas"
       -DCMAKE_Fortran_COMPILER=${lib.getDev mpi}/bin/mpif90
-      ${lib.optionalString passthru.isILP64 ''
-        -DCMAKE_Fortran_FLAGS="-fdefault-integer-8"
-        -DCMAKE_C_FLAGS="-DInt=long"
-      ''}
+      -DCMAKE_C_FLAGS="${lib.concatStringsSep " " [
+            "-Wno-implicit-function-declaration"
+            (lib.optionalString passthru.isILP64 "-DInt=long")
+      ]}"
+      ${lib.optionalString passthru.isILP64 ''-DCMAKE_Fortran_FLAGS="-fdefault-integer-8"''}
       )
   '';
 

